docs/source/en/quantization/torchao.md

```diff
@@ -0,0 +1,92 @@
+<!-- Copyright 2024 The HuggingFace Team. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+the License. You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+specific language governing permissions and limitations under the License. -->
+
+# torchao
+
+[TorchAO](https://github.com/pytorch/ao) is an architecture optimization library for PyTorch. It provides high-performance dtypes, optimization techniques, and kernels for inference and training, featuring composability with native PyTorch features like [torch.compile](https://pytorch.org/tutorials/intermediate/torch_compile_tutorial.html), FullyShardedDataParallel (FSDP), and more.
+
+Before you begin, make sure you have Pytorch 2.5+ and TorchAO installed.
```

Comment:
Indeed it seems Pytorch 2.5+ is required because in https://github.com/huggingface/diffusers/blob/92933ec36a13989981a6fc4189857e8b4dc2c38d/src/diffusers/quantizers/torchao/torchao_quantizer.py#L45 there is an import of `torch.uint1` (and others) which are not available in earlier torch versions. However, `diffusers` seem to require `torch>=1.4` ([ref](https://github.com/huggingface/diffusers/blob/4b557132ce955d58fd84572c03e79f43bdc91450/setup.py#L133)), so this seem inconsistent. Am I missing something?

Reply:
TorchAO will not be imported or usable unless the pytorch version of 2.5 or above is available. Some Diffusers models can run with the 1.4 version as well, which is why that's the minimum required version.

Reply:
I'm running into the same issue with the `troch.unit1` import. It seems the TorchAO import is not guarded according to the backtrace. The following backtrace stems from this import line:

```
from diffusers import StableDiffusionXLPipeline
```

Here is the trace, and the pip list:

```
    Traceback (most recent call last):
      File "/github/home/.local/lib/python3.10/site-packages/diffusers/utils/import_utils.py", line 920, in _get_module
        return importlib.import_module("." + module_name, self.__name__)
      File "/usr/local/lib/python3.10/importlib/__init__.py", line 126, in import_module
        return _bootstrap._gcd_import(name[level:], package, level)
      File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
      File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
      File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
      File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
      File "<frozen importlib._bootstrap_external>", line 883, in exec_module
      File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
      File "/github/home/.local/lib/python3.10/site-packages/diffusers/loaders/single_file.py", line 24, in <module>
        from .single_file_utils import (
      File "/github/home/.local/lib/python3.10/site-packages/diffusers/loaders/single_file_utils.py", line 28, in <module>
        from ..models.modeling_utils import load_state_dict
      File "/github/home/.local/lib/python3.10/site-packages/diffusers/models/modeling_utils.py", line 35, in <module>
        from ..quantizers import DiffusersAutoQuantizer, DiffusersQuantizer
      File "/github/home/.local/lib/python3.10/site-packages/diffusers/quantizers/__init__.py", line 15, in <module>
        from .auto import DiffusersAutoQuantizer
      File "/github/home/.local/lib/python3.10/site-packages/diffusers/quantizers/auto.py", line 31, in <module>
        from .torchao import TorchAoHfQuantizer
      File "/github/home/.local/lib/python3.10/site-packages/diffusers/quantizers/torchao/__init__.py", line 15, in <module>
        from .torchao_quantizer import TorchAoHfQuantizer
      File "/github/home/.local/lib/python3.10/site-packages/diffusers/quantizers/torchao/torchao_quantizer.py", line 45, in <module>
        torch.uint1,
      File "/github/home/.local/lib/python3.10/site-packages/torch/__init__.py", line 1938, in __getattr__
        raise AttributeError(f"module '{__name__}' has no attribute '{name}'")
    AttributeError: module 'torch' has no attribute 'uint1'
```

And the pip list:

<details>
<summary>pip list -v</summary>

```
Package                  Version     Editable project location    Location                                         Installer
------------------------ ----------- ---------------------------- ------------------------------------------------ ---------
certifi                  2024.12.14                               /github/home/.local/lib/python3.10/site-packages pip
charset-normalizer       3.4.0                                    /github/home/.local/lib/python3.10/site-packages pip
colorama                 0.4.6                                    /github/home/.local/lib/python3.10/site-packages pip
coloredlogs              15.0.1                                   /github/home/.local/lib/python3.10/site-packages pip
colorlog                 6.9.0                                    /github/home/.local/lib/python3.10/site-packages pip
coverage                 7.6.9                                    /github/home/.local/lib/python3.10/site-packages pip
diffusers                0.32.0                                   /github/home/.local/lib/python3.10/site-packages pip
exceptiongroup           1.2.2                                    /github/home/.local/lib/python3.10/site-packages pip
execnet                  2.1.1                                    /github/home/.local/lib/python3.10/site-packages pip
filelock                 3.16.1                                   /github/home/.local/lib/python3.10/site-packages pip
flatbuffers              24.12.23                                 /github/home/.local/lib/python3.10/site-packages pip
fsspec                   2024.12.0                                /github/home/.local/lib/python3.10/site-packages pip
huggingface-hub          0.27.0                                   /github/home/.local/lib/python3.10/site-packages pip
humanfriendly            10.0                                     /github/home/.local/lib/python3.10/site-packages pip
idna                     3.10                                     /github/home/.local/lib/python3.10/site-packages pip
importlib_metadata       [8](https://github.qualcomm.com/Hexagon-Architecture/piqaro/actions/runs/952286/job/1969259#step:10:9).5.0                                    /github/home/.local/lib/python3.10/site-packages pip
iniconfig                2.0.0                                    /github/home/.local/lib/python3.10/site-packages pip
Jinja2                   3.1.5                                    /github/home/.local/lib/python3.10/site-packages pip
markdown-it-py           3.0.0                                    /github/home/.local/lib/python3.10/site-packages pip
MarkupSafe               3.0.2                                    /github/home/.local/lib/python3.10/site-packages pip
mdurl                    0.1.2                                    /github/home/.local/lib/python3.10/site-packages pip
mpmath                   1.3.0                                    /github/home/.local/lib/python3.10/site-packages pip
networkx                 3.4.2                                    /github/home/.local/lib/python3.10/site-packages pip
numpy                    1.26.4                                   /github/home/.local/lib/python3.10/site-packages pip
nvidia-cublas-cu12       12.1.3.1                                 /github/home/.local/lib/python3.10/site-packages pip
nvidia-cuda-cupti-cu12   12.1.105                                 /github/home/.local/lib/python3.10/site-packages pip
nvidia-cuda-nvrtc-cu12   12.1.105                                 /github/home/.local/lib/python3.10/site-packages pip
nvidia-cuda-runtime-cu12 12.1.105                                 /github/home/.local/lib/python3.10/site-packages pip
nvidia-cudnn-cu12        8.[9](https://github.qualcomm.com/Hexagon-Architecture/piqaro/actions/runs/952286/job/1969259#step:10:10).2.26                                 /github/home/.local/lib/python3.10/site-packages pip
nvidia-cufft-cu12        11.0.2.54                                /github/home/.local/lib/python3.[10](https://github.qualcomm.com/Hexagon-Architecture/piqaro/actions/runs/952286/job/1969259#step:10:11)/site-packages pip
nvidia-curand-cu12       10.3.2.106                               /github/home/.local/lib/python3.10/site-packages pip
nvidia-cusolver-cu12     [11](https://github.qualcomm.com/Hexagon-Architecture/piqaro/actions/runs/952286/job/1969259#step:10:12).4.5.107                               /github/home/.local/lib/python3.10/site-packages pip
nvidia-cusparse-cu[12](https://github.qualcomm.com/Hexagon-Architecture/piqaro/actions/runs/952286/job/1969259#step:10:13)     12.1.0.106                               /github/home/.local/lib/python3.10/site-packages pip
nvidia-nccl-cu12         2.19.3                                   /github/home/.local/lib/python3.10/site-packages pip
nvidia-nvjitlink-cu12    12.6.85                                  /github/home/.local/lib/python3.10/site-packages pip
nvidia-nvtx-cu12         12.1.105                                 /github/home/.local/lib/python3.10/site-packages pip
onnx                     1.17.0                                   /github/home/.local/lib/python3.10/site-packages pip
onnx2torch               1.5.15                                   /github/home/.local/lib/python3.10/site-packages pip
onnxruntime              1.20.1                                   /github/home/.local/lib/python3.10/site-packages pip
onnxsim                  0.4.36                                   /github/home/.local/lib/python3.10/site-packages pip
packaging                24.2                                     /github/home/.local/lib/python3.10/site-packages pip
pandas                   2.2.3                                    /github/home/.local/lib/python3.10/site-packages pip
pillow                   11.0.0                                   /github/home/.local/lib/python3.10/site-packages pip
pip                      22.0.4                                   /usr/local/lib/python3.10/site-packages          pip
pluggy                   1.5.0                                    /github/home/.local/lib/python3.10/site-packages pip
protobuf                 5.29.2                                   /github/home/.local/lib/python3.10/site-packages pip
Pygments                 2.18.0                                   /github/home/.local/lib/python3.10/site-packages pip
pytest                   8.3.4                                    /github/home/.local/lib/python3.10/site-packages pip
pytest-xdist             3.6.1                                    /github/home/.local/lib/python3.10/site-packages pip
python-dateutil          2.9.0.post0                              /github/home/.local/lib/python3.10/site-packages pip
python-dotenv            1.0.1                                    /github/home/.local/lib/python3.10/site-packages pip
python-json-logger       3.2.1                                    /github/home/.local/lib/python3.10/site-packages pip
pytz                     2024.2                                   /github/home/.local/lib/python3.10/site-packages pip
PyYAML                   6.0.2                                    /github/home/.local/lib/python3.10/site-packages pip
regex                    2024.11.6                                /github/home/.local/lib/python3.10/site-packages pip
requests                 2.32.3                                   /github/home/.local/lib/python3.10/site-packages pip
rich                     13.9.4                                   /github/home/.local/lib/python3.10/site-packages pip
safetensors              0.4.5                                    /github/home/.local/lib/python3.10/site-packages pip
scipy                    1.[14](https://github.qualcomm.com/Hexagon-Architecture/piqaro/actions/runs/952286/job/1969259#step:10:15).1                                   /github/home/.local/lib/python3.10/site-packages pip
sentencepiece            0.2.0                                    /github/home/.local/lib/python3.10/site-packages pip
setuptools               58.1.0                                   /usr/local/lib/python3.10/site-packages          pip
six                      1.17.0                                   /github/home/.local/lib/python3.10/site-packages pip
sympy                    1.13.3                                   /github/home/.local/lib/python3.10/site-packages pip
tabulate                 0.9.0                                    /github/home/.local/lib/python3.10/site-packages pip
tokenizers               0.[15](https://github.qualcomm.com/Hexagon-Architecture/piqaro/actions/runs/952286/job/1969259#step:10:16).2                                   /github/home/.local/lib/python3.10/site-packages pip
tomli                    2.2.1                                    /github/home/.local/lib/python3.10/site-packages pip
torch                    2.2.2                                    /github/home/.local/lib/python3.10/site-packages pip
torchvision              0.[17](https://github.qualcomm.com/Hexagon-Architecture/piqaro/actions/runs/952286/job/1969259#step:10:18).2                                   /github/home/.local/lib/python3.10/site-packages pip
tqdm                     4.67.1                                   /github/home/.local/lib/python3.10/site-packages pip
transformers             4.38.2                                   /github/home/.local/lib/python3.10/site-packages pip
triton                   2.2.0                                    /github/home/.local/lib/python3.10/site-packages pip
typing_extensions        4.12.2                                   /github/home/.local/lib/python3.10/site-packages pip
tzdata                   [20](https://github.qualcomm.com/Hexagon-Architecture/piqaro/actions/runs/952286/job/1969259#step:10:21)24.2                                   /github/home/.local/lib/python3.10/site-packages pip
urllib3                  2.3.0                                    /github/home/.local/lib/python3.10/site-packages pip
wheel                    0.37.1                                   /usr/local/lib/python3.10/site-packages          pip
zipp                     3.[21](https://github.qualcomm.com/Hexagon-Architecture/piqaro/actions/runs/952286/job/1969259#step:10:22).0                                   /github/home/.local/lib/python3.10/site-packages pip
```

</details>

Reply:
Thanks a lot for reporting @fjeremic! We were able to replicate for torch <= 2.2. It seems to not cause the import errors for >= 2.3. We will be doing a patch release soon to fix this behaviour. Sorry for the inconvenience!

Reply:
Thanks for providing a quick fix! 
For completeness, I was running into the import error with `torch 2.2.2` when importing `AutoencoderKL`

Reply:
@BeckerFelix @fjeremic The patch release is out! Hope it fixes any problems you were facing in torch < 2.3